crates/stages/src/stages/bodies.rs

```diff
@@ -125,6 +127,13 @@ impl<DB: Database, D: BodyDownloader> Stage<DB> for BodyStage<D> {
                         },
                     )?;
 
+                    // write transaction block index
+                    if !block.body.is_empty() {
+                        // -1 is here as current_tx_id points to the next transaction.
+                        tx_block_cursor
+                            .append(current_tx_id - 1 + block.body.len() as u64, block.number)?;
```

Comment:
Is this correct?

Assume:
- Block 0: First tx ID is 0, the block has 10 transactions (IDs: 0..=9)
- Block 1: First tx ID is 10, the block has 1 transaction (IDs: 10)

The table will look like this:

- `(10, 0)`
- `(11, 1)`

Checking the implementation of `transaction_block` further down, looking up the transaction with ID 10 would give us position the cursor at the first entry where they key is greater than or equal to 10 (`MDBX_SET_RANGE`), which would give us block 0.

If the table was insted:

- `(9, 0)`
- `(10, 1)`

It would be correct.

So unless I am doing the math wrong, this should be:

```rs
// -1 as `current_tx_id` points to the next transaction
let first_tx_id = current_tx_id - 1;
let highest_tx_id = first_tx_id + block.body.len() - 1;
```

However this needs to handle the case where `block.body.is_empty()` - in that case, `highest_tx_id = first_tx_id` (and the table would still work).

Reply:
> Is this correct?
> 
> Assume:
> 
> * Block 0: First tx ID is 0, the block has 10 transactions (IDs: 0..=9)
> * Block 1: First tx ID is 10, the block has 1 transaction (IDs: 10)
> 
> The table will look like this:
> 
> * `(10, 0)`
> * `(11, 1)`

This is how it was done, table key would point to last tx num of the block.
* `(9,0)`
* `(10,1)`

> 
> Checking the implementation of `transaction_block` further down, looking up the transaction with ID 10 would give us position the cursor at the first entry where they key is greater than or equal to 10 (`MDBX_SET_RANGE`), which would give us block 0.
> 
> If the table was insted:
> 
> * `(9, 0)`
> * `(10, 1)`
> 
> It would be correct.

I did implement it like this. 

> 
> So unless I am doing the math wrong, this should be:
> 
> ```rust
> // -1 as `current_tx_id` points to the next transaction
> let first_tx_id = current_tx_id - 1;
This assumption is not correct. This line was before transaction was inserted to so current_tx_id is pointing to the next tx that is going to be inserted

> let highest_tx_id = first_tx_id + block.body.len() - 1;
> ```
Yeah, in code it would be more correct to write it as: `current_tx_id  + block.body.len() as u64 - 1`
 
I rewrote this part and force committed without first reading this, sorry about that, it now used body.last_tx_id() so it is easier to see intention:
```
if !body.is_empty() {
       tx_block_cursor.append(body.last_tx_index(), block.number)?;
}
```
 
> However this needs to handle the case where `block.body.is_empty()` - in that case, `highest_tx_id = first_tx_id` (and the table would still work).

Empty body number is not found as this is last tx_num to block number relation, as there is no transactions there is not key (TxNumber) for entry to exist.